src/main/kotlin/io/libp2p/core/multiformats/Protocol.kt

```diff
@@ -166,3 +103,100 @@ enum class Protocol(val code: Int, val size: Int, val typeName: String) {
 }
+
+private const val LENGTH_PREFIXED_VAR_SIZE = -1
+
+private val SIZE_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    if (!protocol.hasValue && bytes != null)
+        throw IllegalArgumentException("No value expected for protocol $protocol, but got ${bytes.contentToString()}")
+    if (protocol.hasValue) {
+        requireNotNull(bytes) { "Non-null value expected for protocol $protocol" }
+        if (protocol.sizeBits != LENGTH_PREFIXED_VAR_SIZE && bytes.size * 8 != protocol.sizeBits) {
+            throw IllegalArgumentException("Value of size ${protocol.sizeBits / 8} expected for protocol $protocol but got ${bytes.contentToString()}")
+        }
+    }
+}
+private val NON_EMPTY_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    requireNotNull(bytes) { "Non-null value expected for protocol $protocol" }
+    require(bytes.isNotEmpty()) { "Non-empty value expected for protocol $protocol" }
+}
+private val PEER_ID_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    if (bytes == null)
+        throw IllegalArgumentException("Non-null value expected for PeerId in $protocol")
+    PeerId(bytes) // constructor validates array size
+}
+
+private val NO_PARSER: (Protocol, String) -> ByteArray = { protocol, _ ->
+    throw IllegalArgumentException("No value serializer for protocol $protocol")
+}
+private val NO_DESERIALIZER: (Protocol, ByteArray) -> String = { protocol, _ ->
+    throw IllegalArgumentException("No value deserializer for protocol $protocol")
+}
+private val IP4_PARSER: (Protocol, String) -> ByteArray = { _, addr ->
+    if (!InetAddresses.isInetAddress(addr)) {
+        throw IllegalArgumentException("Couldn't parse IPv4 IP: `$addr`")
+    }
+    val inetAddr = Inet4Address.getByName(addr)
+    if (inetAddr !is Inet4Address) {
+        throw IllegalArgumentException("The address is not IPv4 address: $addr")
+    }
+    inetAddr.address
+}
+private val IP4_DESERIALIZER: (Protocol, ByteArray) -> String = { _, bytes ->
+    Inet4Address.getByAddress(bytes).toString().drop(1)
+}
+private val IP6_PARSER: (Protocol, String) -> ByteArray = { _, addr ->
+    Inet6Address.getByName(addr).address
+}
+private val IP6_DESERIALIZER: (Protocol, ByteArray) -> String = { _, bytes ->
+    Inet6Address.getByAddress(bytes).toString().drop(1)
+}
+private val UINT16_PARSER: (Protocol, String) -> ByteArray = { _, addr ->
+    val x = Integer.parseInt(addr)
+    if (x < 0 || x > 65535) throw IllegalArgumentException("Failed to parse $addr value (expected 0 <= $x < 65536")
+    byteBuf(2).writeShort(x).toByteArray()
+}
+private val UINT16_DESERIALIZER: (Protocol, ByteArray) -> String = { _, bytes ->
+    bytes.toByteBuf().readUnsignedShort().toString()
+}
+private val UTF8_PARSER: (Protocol, String) -> ByteArray = { _, addr ->
+    addr.toByteArray(StandardCharsets.UTF_8)
+}
+private val UTF8_DESERIALIZER: (Protocol, ByteArray) -> String = { _, bytes ->
+    String(bytes, StandardCharsets.UTF_8)
```

Comment:
I don't think this will reject invalid characters (just replace them with `?`).  If we want to fully validate we could use `StandardCharsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(CodingErrorAction.REPORT).decode()` but that adds a bunch of complexity and I'm not entirely sure it's worth it.

Reply:
Right, better be safe. However I've added a UTF-8 validator instead: 6375f2a